### PR TITLE
amd64 UEFI loader copy_staging work and msdosfs_rename reports

### PR DESCRIPTION
**Pull request by kostikbel**

### CODE REVIEW

File: 2021q3/copy_staging.adoc [hunk 1]
```diff
@@ -0,0 +1,76 @@
+== Changes to UEFI boot on amd64 staging
```
Comment (emaste) [inline 2021q3/copy_staging.adoc:1]: I would drop "staging" here, it refers to an aspect of the boot but reads as if this is part of an "amd64 staging" branch.

How about simply "Improved amd64 UEFI boot"?

File: 2021q3/copy_staging.adoc [hunk 2]
```diff
@@ -0,0 +1,76 @@
+== Changes to UEFI boot on amd64 staging 
+
+Contact: Konstantin Belousov <kib@FreeBSD.org>  
+
+UEFI BIOS on PC provides much more rich and streamlined environment
+for pre-OS programs, but also it imposes some requirements on the
+programs that are executed there, OS loaders in particular.  One of
+such requirement is that loader must coordinate its memory use with
```
Comment (emaste) [inline 2021q3/copy_staging.adoc:8]: One such requirement (no "of")

File: 2021q3/copy_staging.adoc [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+BIOS, only using memory that was allocated for it.  Even after loader
+handoff to the operating system kernel, there are still some memory
+regions that are reserved by BIOS for different reasons, most knows
```
Comment (emaste) [inline 2021q3/copy_staging.adoc:11]: ... different reasons.
Examples are runtime...

File: 2021q3/copy_staging.adoc [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+are runtime services code and data.
+
+On the other hand, legacy/CSM BIOS boot works with memory completely
+differently, there are known memory regions that hold BIOS data and
+must be avoided.  Otherwise, the memory is considered free for loader
+and OS to use. [Of course it is not that straightforward, the
```
Comment (emaste) [inline 2021q3/copy_staging.adoc:17]: () are typical, (Of course ... workarounds there.)

File: 2021q3/copy_staging.adoc [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+definition of known regions is up to the vendor and there are a lot of
+workarounds there]
+
+BIOS boot puts the kernel and preloaded data (like modules, memory
+disk, CPU microcode update etc) at the contigous physical memory block
+starting at 2M.  This algorithm goes back to how i386 kernel boots.
+
+Also, when preparing to handle the control of to the kernel, loader
```
Comment (emaste) [inline 2021q3/copy_staging.adoc:25]: Maybe

... preparing to pass control to the kernel, the loader creates...

File: 2021q3/copy_staging.adoc [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+creates quite special temporal mappings, where low 1G of physical
+address space is mapped 1:1 into virtual address space, and then
+repeated for each 1G until the virtual memory end.  Kernel knows about
+its physical location and the temporal mapping, and constructs kernel
+page tables assuming that the physical address of the text is at 2M.
+
+This mechanism of loader to kernel handoff was left unchanged when
+loader got support for UEFI environment.  Loader prepares kernel and
```
Comment (emaste) [inline 2021q3/copy_staging.adoc:33]: maybe "loader gained support for the UEFI environment"

File: 2021q3/copy_staging.adoc [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+auxillary preload data in so called staging area while UEFI boot
```
Comment (emaste) [inline 2021q3/copy_staging.adoc:34]: in //a// so called staging

File: 2021q3/copy_staging.adoc [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+services are active, and after EFI_BOOT_SERVICES.ExitBootServices(),
+the temporal mapping is activated and staging area is copied at 2M.
+
+An advantage at that time was that no changes to the kernel were
+needed.  But there are issues, the biggest is that memory at 2M might
+be not free for reuse.  For instance, BIOS runtime code or data might
+be located there.  Or there might be no memory at 2M at all.  Or
+trampoling page table or code, or even some parts of the staging area
+overlapped with the 2M region where staging area is copied.  The
+outcome was a weird and hard to diagnose boot time failures, typically
```
Comment (emaste) [inline 2021q3/copy_staging.adoc:44]: "a" doesn't match "failures"
maybe
"hard to diagnose boot time failure, typically a hard hang when the loader started the kernel."

File: 2021q3/copy_staging.adoc [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+machine hangs hard when loader started kernel.
+
+Another limitation is the 1G transient mapping, which due to copying
+means that total size of preloaded data cannot exceed around 400M for
+everything, including kernel, memory disks, and anything else.  Also
+the code to grow staging area on demand was quite unflexible, only
+able to grow the staging area in place.
+
+Described work allows UEFI loader on amd64 to start kernel from the
```
Comment (emaste) [inline 2021q3/copy_staging.adoc:53]: The work described in this report allows
or just
This project allows

File: 2021q3/copy_staging.adoc [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+staging area without copying.  Kernel assumptions about the hand-off
+were explicitly identified and documented.  Kernel only requires that
+the staging area was located below 4G together with the trampoline
+page table (this is a consequence of CPU architecture requiring 32bit
+protected mode to enter long mode), be 2M aligned and whole low 4G
+mapped 1:1 at hand-off.  Kernel deduces its physical address and
+builds kernel page tables accordingly.
+
+Making booting with staging area in-place work required identifying
+all places where amd64 kernel grown a dependency on its physical
```
Comment (emaste) [inline 2021q3/copy_staging.adoc:63]: grew a dependency
or perhaps
had a dependency

File: 2021q3/copy_staging.adoc [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+location.  Most complicated part was application processors startup,
+which required rewriting initialization code, which we were able to
+streamline as result.  In particular, when AP enters paging mode, it
+does so right into correct kernel page table, without loading
+intermediate trampoline page table.
+
+Updated loader automatically detects if the loaded kernel can handle
+in-place staging are ('non-copying mode').  If needed, this can be
```
Comment (emaste) [inline 2021q3/copy_staging.adoc:71]: staging //area// I presume?

File: 2021q3/copy_staging.adoc [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+controlled by loader copy_staging command.  Also, the code to grow
```
Comment (emaste) [inline 2021q3/copy_staging.adoc:72]: maybe "overridden with the ... command". Perhaps include an inline example if it doesn't make this too awkward.

File: 2021q3/msdosfs_rename.adoc [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+Recently trasz@ found that msdosfs rename can be easily deadlocked.
+Further code examination revealed a lot of issues with locking,
+potential use after free, and filesystem structure corruption.
+
+During this work, all known issues were fixed and msdosfs passes
+enchanced stress2 suite of tests.
+
+In particular, locking in the msdosfs rename code was reworked: we
+need to lock up to four vnodes, and check one path to ensure that
+rename does not create cyclic parent relations between directories.
```
Comment (pauamma) [inline 2021q3/msdosfs_rename.adoc:20]: I may be overly picky here, but I'd s/cyclic/circular/.

File: 2021q3/msdosfs_rename.adoc [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+For that, the locking procedure was copied from UFS rename, where all
+vnodes except the first are locked in try-mode.  Lockess relockup was
+added to msdosfs and directory path checker changed to non-blocking
```
Comment (pauamma) [inline 2021q3/msdosfs_rename.adoc:23]: s/and/and the/

File: 2021q3/msdosfs_rename.adoc [hunk 15]
```diff
@@ -0,0 +1,26 @@
...
+mode.
+
+Sponsor: The FreeBSD Foundation (kib)
```
Comment (pauamma) [inline 2021q3/msdosfs_rename.adoc:26]: Maybe elaborate what "(kib)" means? Maybe "(work by kib)"?

File: 2021q3/copy_staging.adoc [hunk 16]
```diff
@@ -0,0 +1,78 @@
+== Improved amd64 UEFI boot
+
+Contact: Konstantin Belousov <kib@FreeBSD.org>  
+
+UEFI BIOS on PC provides much more rich and streamlined environment
```
[the remark this answers is not in the document]
Comment (emaste) [reply 2021q3/copy_staging.adoc:5]: streamlined would have to change too then, "a much richer and more streamlined ..."

File: 2021q3/copy_staging.adoc [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+for pre-OS programs, but also it imposes some requirements on the
+programs that are executed there, OS loaders in particular.  One
+such requirement is that loader must coordinate its memory use with
+BIOS, only using memory that was allocated for it.  Even after loader
```
[the remark this answers is not in the document]
Comment (emaste) [reply 2021q3/copy_staging.adoc:9]: yeah, I think "the BIOS" here sounds better

File: 2021q3/copy_staging.adoc [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+and OS to use. (Of course it is not that straightforward, the
+definition of known regions is up to the vendor and there are a lot of
+workarounds there.)
+
+BIOS boot puts the kernel and preloaded data (like modules, memory
+disk, CPU microcode update etc) at the contigous physical memory block
+starting at 2M.  This algorithm goes back to how i386 kernel boots.
+
+Also, when preparing to pass control to the kernel, the loader
+creates quite special temporal mappings, where low 1G of physical
```
[the remark this answers is not in the document]
Comment (emaste) [reply 2021q3/copy_staging.adoc:26]: "very" seems slightly more typical to me, although for quarterly report entries I think it isn't critical

File: 2021q3/msdosfs_rename.adoc [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+locking and miscellaneous bugfixes over time, sometimes a serious issue
+is found in it.
+
+Recently trasz@ found that msdosfs rename can be easily deadlocked.
+Further examination of rename code revealed a lot of issues with locking,
+potential use after free, and filesystem structure corruption.
+
+As part of the update, locking in the msdosfs rename code was reworked.
+We need to lock up to four vnodes, and check one path to ensure that
+rename does not create cicrular parent relations between directories.
```
Comment (pauamma) [inline 2021q3/msdosfs_rename.adoc:17]: s/cicrular/circular/

File: 2021q3/copy_staging.adoc [hunk 20]
```diff
@@ -0,0 +1,78 @@
+== Improved amd64 UEFI boot
+
+Contact: Konstantin Belousov <kib@FreeBSD.org>  
+
+UEFI BIOS on PC provides much more rich and streamlined environment
+for pre-OS programs, but also it imposes some requirements on the
+programs that are executed there, OS loaders in particular.  One
+such requirement is that loader must coordinate its memory use with
```
[the remark this answers is not in the document]
Comment (pauamma) [reply 2021q3/copy_staging.adoc:8]: Still needs "the", I think.

File: 2021q3/copy_staging.adoc [hunk 21]
```diff
@@ -0,0 +1,78 @@
...
+
+Also, when preparing to pass control to the kernel, the loader
+creates very special temporary mappings, where low 1G of physical
+address space is mapped 1:1 into virtual address space, and then
+repeated for each 1G until the virtual memory end.  The kernel knows about
+its physical location and the temporary mapping, and constructs kernel
+page tables assuming that the physical address of the text is at 2M.
+
+This mechanism of loader to kernel handoff was left unchanged when
+the loader gained support for the UEFI environment.  The oader prepares kernel and
```
Comment (pauamma) [inline 2021q3/copy_staging.adoc:33]: s/The oader/The loader/

File: 2021q3/copy_staging.adoc [hunk 22]
```diff
@@ -0,0 +1,78 @@
...
+auxiliary preload data in so called staging area while UEFI boot
```
Comment (pauamma) [inline 2021q3/copy_staging.adoc:34]: s/so called/so-called/

Also needs "the" or "a" (not sure which) after "in".

File: 2021q3/copy_staging.adoc [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+services are active, and after EFI_BOOT_SERVICES.ExitBootServices(),
+the temporary mapping is activated and the staging area is copied at 2M.
+
+An advantage at that time was that no changes to the kernel were
+needed.  But there are issues, the biggest is that memory at 2M might
```
Comment (pauamma) [inline 2021q3/copy_staging.adoc:39]: s/issues,/issues;/ (avoid comma splice)

File: 2021q3/copy_staging.adoc [hunk 24]
```diff
@@ -0,0 +1,78 @@
...
+outcome was a hard to diagnose boot time failure, typically a hard hang
+when the loader started the kernel.
+
+Another limitation is the 1G transient mapping, which due to copying
+means that the total size of preloaded data cannot exceed around 400M for
+everything, including kernel, memory disks, and anything else.  Also
+the code to grow the staging area on demand was quite unflexible, only
+able to grow the staging area in place.
+
+The work described in this report allows UEFI loader on amd64 to start
```
Comment (pauamma) [inline 2021q3/copy_staging.adoc:53]: s/UEFI/the UEFI/

File: 2021q3/copy_staging.adoc [hunk 25]
```diff
@@ -0,0 +1,78 @@
...
+streamline as result.  In particular, when an AP enters paging mode, it
+does so straight into correct kernel page table, without loading
+intermediate trampoline page table.
+
+The updated loader automatically detects if the loaded kernel can handle
+in-place staging area ('non-copying mode').  If needed, this can be
+overridden with the loader's copy_staging command.  For instance,
+'copy_staging enable' tells the loader to unconditionally copy the staging
+area to 2M regardless of kernel capabilities (default is 'copy_staging auto').
+Also, the code to grow staging area was made much more robust,
```
Comment (pauamma) [inline 2021q3/copy_staging.adoc:75]: s/the staging area/staging area/

File: 2021q3/copy_staging.adoc [hunk 26]
```diff
@@ -0,0 +1,78 @@
...
+requiring 32bit protected mode to enter long mode), be 2M aligned and
+the whole low 4G mapped 1:1 at hand-off.  The kernel computes its physical
+address and builds kernel page tables accordingly.
+
+Making the kernel boot with staging area in-place required identifying
+all places where the amd64 kernel had a dependency on its physical
+location.  Most complicated part was application processors startup,
+which required rewriting initialization code, which we were able to
+streamline as result.  In particular, when an AP enters paging mode, it
+does so straight into correct kernel page table, without loading
```
Comment (pauamma) [inline 2021q3/copy_staging.adoc:67]: s/correct/the correct/

File: 2021q3/copy_staging.adoc [hunk 27]
```diff
@@ -0,0 +1,78 @@
...
+about the hand-off were explicitly identified and documented.  The kernel
+only requires the staging area to be located below 4G together with
+the trampoline page table (this is a consequence of CPU architecture
+requiring 32bit protected mode to enter long mode), be 2M aligned and
+the whole low 4G mapped 1:1 at hand-off.  The kernel computes its physical
+address and builds kernel page tables accordingly.
+
+Making the kernel boot with staging area in-place required identifying
+all places where the amd64 kernel had a dependency on its physical
+location.  Most complicated part was application processors startup,
```
Comment (pauamma) [inline 2021q3/copy_staging.adoc:64]: s/Most/The most/

File: 2021q3/msdosfs_rename.adoc [hunk 28]
```diff
@@ -0,0 +1,26 @@
...
+rename does not create circular parent relations between directories.
+For that, the locking procedure was copied from UFS rename, where all
+vnodes except the first are locked in try-mode.  Lockless relockup was
+added to msdosfs and the directory path checker was changed to non-blocking
+mode.
+
+During this work, all known issues were fixed and msdosfs passes
+enhanced stress2 suite of tests.
+
+Sponsor: The FreeBSD Foundation (kib' contributions)
```
Comment (pauamma) [inline 2021q3/msdosfs_rename.adoc:26]: s/kib/kib's/ ("not ending in s or z" is the rule, IIRC)

**Review by pauamma — CHANGES_REQUESTED** [review]

**Review by pauamma — APPROVED** [review]
LGTM.

**Pull request merged**